internal/controller/referencegrant.go

```diff
@@ -0,0 +1,190 @@
+// Copyright Envoy AI Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwapiv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	aigv1a1 "github.com/envoyproxy/ai-gateway/api/v1alpha1"
+)
+
+const (
+	// aiServiceBackendGroup is the API group for AIServiceBackend.
+	aiServiceBackendGroup = "aigateway.envoyproxy.io"
+	// aiServiceBackendKind is the kind for AIServiceBackend.
+	aiServiceBackendKind = "AIServiceBackend"
+	// aiGatewayRouteKind is the kind for AIGatewayRoute.
+	aiGatewayRouteKind = "AIGatewayRoute"
+)
+
+// ReferenceGrantValidator validates cross-namespace references using ReferenceGrant resources.
+type ReferenceGrantValidator struct {
+	client client.Client
+}
+
+// NewReferenceGrantValidator creates a new ReferenceGrantValidator.
+func NewReferenceGrantValidator(c client.Client) *ReferenceGrantValidator {
+	return &ReferenceGrantValidator{client: c}
+}
+
+// ValidateAIServiceBackendReference validates that an AIGatewayRoute can reference an AIServiceBackend
+// in a different namespace by checking for a valid ReferenceGrant.
+//
+// Parameters:
+//   - ctx: context for the operation
+//   - routeNamespace: namespace of the AIGatewayRoute
+//   - backendNamespace: namespace of the AIServiceBackend
+//   - backendName: name of the AIServiceBackend (optional, for logging)
+//
+// Returns:
+//   - error: nil if the reference is valid (same namespace or valid ReferenceGrant exists), error otherwise
+func (v *ReferenceGrantValidator) ValidateAIServiceBackendReference(
+	ctx context.Context,
+	routeNamespace string,
+	backendNamespace string,
+	backendName string,
+) error {
+	// Same namespace references don't need ReferenceGrant
+	if routeNamespace == backendNamespace {
+		return nil
+	}
+
+	// List all ReferenceGrants in the backend namespace
+	var referenceGrants gwapiv1b1.ReferenceGrantList
+	if err := v.client.List(ctx, &referenceGrants, client.InNamespace(backendNamespace)); err != nil {
+		return fmt.Errorf("failed to list ReferenceGrants in namespace %s: %w", backendNamespace, err)
+	}
+
+	// Check if any ReferenceGrant allows this cross-namespace reference
+	for _, grant := range referenceGrants.Items {
```

Comment:
nit but let's avoid the copy unnecessarily 
```suggestion
	for i  := range referenceGrants.Items {
		grant := &referenceGrants.Items[i]
```

Reply:
fixed